### DISCUSSION TITLE
Please re-open issue no.345.... (Eco mode switching issue) I have an idea

### DISCUSSION THREAD

**Discussion by joe-700** [seerge/g-helper#834] [Ideas]
Hi Seerge,

hope you are well, sorry for putting this into a discussion, not sure how else to deal with it.

The above issue was closed because we suspect that this was down to Windows/Bios/Nvidia behaviour that can't be influenced by the G-Helper application, i.e. out of scope of the application, however this issue is still driving me round the bend so I have a simple solution suggestion that will fix the problem for all users that can be implemented by G-Helper:

**_Simplified Issue Description:_**
Simply put: the system ties itself in a knot when switching to 'Standard' mode from 'Eco' (only when shutdown/booted into 'Eco' mode). From my extensive testing all the reported issues from no.345 stem from that initial point of failure.

**_Solution:_**
Can the G-Helper add a toggle in the 'Extras' GUI that just ensures the mode switch is set to 'Standard' before: 
1. Shutdown
2. Hibernate. 

This will correct the issue until Asus/Nvidia/Windows can get to the bottom of it, if indeed they ever do!!!

Sorry for putting this suggestion into a discussion, if you re-open the issue above I can move this to that thread and you can close this discussion.

Keep up the good work,

Kind regards,
Joe

**Comment by seerge** [comment]
@joe-700 hello, i don't think there is a way to prevent windows from shutting down (and especially hibernating cause that often happens during sleep at all) just from the app itself :)

**Reply by joe-700** [reply]
Sorry, I'm actually suggesting that the G-Helper app listens for shutdown and hibernate events and the listener triggers a switch to 'Normal' mode in those events... The listener being written into the G-Helper app.

Unless there are standard interface implementations for windows apps that can be inherited in your code, then you could just use those?

**Reply by joe-700** [reply]
Check this out -> https://www.apriorit.com/dev-blog/413-win-api-shutdown-events

**Reply by joe-700** [reply]
We only need to ensure that the GPU mode is set to 'Standard' when shutdown or hibernation occurs and bish bash bosh, problem solved.

**Reply by seerge** [reply]
@joe-700 how do we know that your specific device is not the only one that needs this? And anyway for hibernation during this most probably won't work (as windows simply won't allow app to run anything)

**Reply by joe-700** [reply]
Ahh, Good question. Its definitely the case for my machine... we could reopen the issue and advise anyone with eco mode switching issues to try a restart from the standard mode to check this across the community.

As for handling the hibernation from sleep, this is a bit tricky but I'm fairly sure that you can add handlers for: shutdown; hibernation and sleep. So just apply the 'switch to normal' for each case, that way if it goes to sleep the mode is already set to normal and voila.

**Reply by joe-700** [reply]
Also looks like the same problem that Noisrevid was having re. issue 451.

**Reply by seerge** [reply]
I mean i'm not sure if shutting down in "standard" mode is the solution that helps others.

And on a sidenote you can't run any code on hibernate anyway https://stackoverflow.com/questions/6012219/how-to-detect-if-windows-is-going-to-hibernate-or-suspend looks like

Long story short : so far it seems that solution you are asking to add is very time consuming (for me), and does't guarantee any outcomes, and i have no idea how many people need that.

**Reply by joe-700** [reply]
No worries, I'll have a look at this tonight... its been a while since I did any CS in anger.

I was thinking to add some extra subscriptions to the Program.cs, line 100, (you are already using the Win32 reference, that might do it), and GPU switching methods below to go into 'Standard' on shutdown and sleep events.

**Reply by joe-700** [reply]
Hi Seerge,

Bit of an update: Got the code in and it works a treat; It ensures 'Normal' (i.e. safe boot mode) is set before: 1.Shutdown, 2.Hybernate, 3.Hybernate during sleep.
I added the following in case you want to incorporate it into the Git repo:

**_Program.cs ( at line 102)_**
        //----------------------------------------------------------------
            SystemEvents.SessionEnding += c_SessionEndedEvent;
        //----------------------------------------------------------------

**_Program.cs ( at line 121 )_**
        //----------------------------------------------------------------
        private static void c_SessionEndedEvent(object sender, SessionEndingEventArgs e)
        {
            gpuControl.SetGPUMode(1);
        }
        //----------------------------------------------------------------

**_and the following to Program.cs (at current line 180)_**
        //----------------------------------------------------------------
        private static void SystemEvents_PowerModeChanged(object sender, PowerModeChangedEventArgs e)
        {
            if (e.Mode.ToString() == "Suspend")
            {
                gpuControl.SetGPUMode(1);
            }           
            if (SystemInformation.PowerStatus.PowerLineStatus == isPlugged) return;
            Logger.WriteLine("Power Mode Changed");
            SetAutoModes(true);
        }
        //----------------------------------------------------------------

**Reply by joe-700** [reply]
I have a more refined solution now that will track and ensure the mode is reset to the last user selection in the background after restart and resume, i.e. shutdown and hibernate. This works really well and completely solved all my system instability issues re. eco mode switching. This also removes the need to force a GPU driver reset when switching to eco. 

If you add me as a contributor I can upload the solution into the main branch.

**Reply by seerge** [reply]
@joe-700 hello, you can do it in a fork and post a PR so i can check what you have added there :) this is how things work on github.

**Reply by joe-700** [reply]
Will do, I'll look into it tonight. 👍

**Reply by seerge** [reply]
Does BIOS 312 fix your GPU issues btw ? I now have G14 4080, and I can't reproduce anything of mentioned here on my device

**Reply by joe-700** [reply]
Mate, that's super interesting... Bit of an update from my side, I have returned my original 4080 model and just got the replacement, the issue is there with the new one as well straight out of the box, same model; GA402XZ. I have the new BIOS 312, was hoping that this would fix it but no luck.

It is very easy to replicate, just set to 'eco', restart and set back to 'standard', you will know straight away because you loose brightness control, (the slider appears but doesn't do anything), I also get a system sound similar to the one I get when I connect an external monitor. Once this happens, I also loose optimus settings in Nvidia control panel and GPU mode switching starts to glitch.

Are you running windows 11?
Do you have the same model number: GA402XZ?

Its a fantastic laptop BTW; besides this issue I have. hope you don't end up with this problem. I can't really use 'eco' or 'optimised' unless  I enable the G-Helper function.

**Reply by seerge** [reply]
Ok, I have finally managed to reproduce it. As it doesn't happen all the time.

It's clearly windows/software issue. As disabling / enabling NVidia device in device manager instanly solves problem. The only thing I have noted that default 3 seconds delay scripted in the app between disable/enable is not always enough 

I have increased it to 5 and it seem to solve issue in 100% of the time for me
[GHelper.zip](https://github.com/seerge/g-helper/files/12304111/GHelper.zip)

As I understand reasons - it's some issue on nvidia driver side, as it locks GPU and refuses to release no matter what (untill windows forces it to)

**Reply by seerge** [reply]
P.S. I just tried to reproduce whole scenario with Armoury Crate. It behaves literally 1 to 1, can't disable GPU after booting in Eco and switching to Standard.

Which proves that all those people on reddit who scream that AC can handle GPU and G-Helper has "bugs" are just wrong ...

**Reply by joe-700** [reply]
The additional delay is not working for me. You are right though, restarting the device seems to bring Nvida options back, however the brightness control is still disabled, is that the same for you?

I have been hounding Asus over this BTW, I can upload the transcripts if your interested, however they haven't been much help yet and continue to insist that I return if for repair, even after I've confirmed the issue on the new machine.

**Reply by joe-700** [reply]
Just seen your last post and I'm smiling away, I have no words... Its just funny... G-Helper is way better than AC, its doing the exact same thing with less nonsense.

**Reply by seerge** [reply]
This is current command app runs (you can manually run it in powershell as admin)

```
$device = Get-PnpDevice | Where-Object { $_.FriendlyName -imatch 'NVIDIA' -and $_.Class -eq 'Display' }; Disable-PnpDevice $device.InstanceId -Confirm:$false; Start-Sleep -Seconds 5; Enable-PnpDevice $device.InstanceId -Confirm:$false
```

Can you check what delay is acceptable for your case (if not 5) ?

-----
Typical example of a complainer https://www.reddit.com/r/ZephyrusG14/comments/15lr395/ghelper_has_too_many_issues/ 

I have a feeling that since AC fails to disable GPU very often, majority of users don't even encounter a case when they shut down laptop in eco mode , and since in g-helper it works - they start noticing it only here

**Reply by seerge** [reply]
I have digested whole thing a bit, and decided to push this "fix" live for all G14 2023 (i.e. models containing GA402)

No need to edit configs anymore, you can just turn this option OFF under Extra

https://github.com/seerge/g-helper/releases/tag/v0.108

**Reply by joe-700** [reply]
Thanks Seerge, 

Agreed,iIf you don't use eco mode then you wont encounter this problem, and when you do its hard to pin down and replicate. Many user will just attribute it to odd software behaviour and blame AC/G-helper. If you use automatic GPU switching for advanced optimus then its actually not possible to select 'eco' mode, another reason many will not even be aware of the problem.

I have a sneaky suspicion that this will be an issue for all G14's (perhaps even all zephyrus and flow laptops) that support advanced optimus.

I will continue to pursue Asus for a fix and escalate where I can. I suggest that other users also raise support tickets with Asus. The more tickets raised the more pressure on them to actually fix it. My advice for those that choose to: reference this thread and refuse to send back for 'repair' (RMA) because we know this will not help and just delay a fix being rolled out.

Thanks for putting the fix into the extras menu. hope you enjoy the new machine. 👌

**Reply by seerge** [reply]
With G-Helper you can force Eco mode from any Optimus state :) (and this is another reason why people just set it more often). But this is intentional and by design.

I think only G14 2023 are affected, I had G16 2023 and G16 Strix 2023 with 4080, both of them have Advanced Optimus but Intel CPUs and didn't had any issues like that. It seem to be Optimus + AMD combo. And in 2023 year it's only G14, X13  and Strix Scars that have AMD CPUs

Most probably Asus won't fix it, as this is more of a Nvidia driver issue. When you boot in Eco - Driver doesn't know about existence of nvidia GPU at all, and then goes nuts. Hopefully Nvidia will fix that eventually
